tools/c7n_azure/c7n_azure/resources/cosmos_db.py

```diff
@@ -355,54 +306,222 @@ def _process_resources(self, resources, event):
+            tag_name = self.data.get('state-tag')
+            container_states_tag_value = TagHelper.get_tag_value(
+                parent_account, tag_name)
+
+            if container_states_tag_value:
+                for state in container_states_tag_value.split(';'):
+                    state_data = state.split(':')
+                    container_rid = state_data[0]
+                    container_throughput = state_data[1]
+
+                    container = next((c for c in resources if c['_rid'] == container_rid), None)
+
+                    if container:
+                        self._process_resource(container, account_client, container_throughput)
+            else:
+                self.log.warning('No tag {} on parent resource, {}.'.format(
+                    tag_name, parent_account))
+
+        except Exception as e:
+            log.warn(e)
+
+        return resources
+
+
+@CosmosDBCollection.action_registry.register('save-throughput-state')
+class CosmosDBStoreStateAction(AzureBaseAction):
+    """CosmosDB Store State Action
+
+    Stores the throughput of collections in a tag on the parent Cosmos DB account.
+    With accounts that have many collections, it's important to filter down which
+    collections to store since there is a tag length limit (approx 16 collections).
+
+    :example:
+
+    This policy stores the throughput of collections with throughput over 400 in
+    a tag called 'on-hour-state' on the parent Cosmos DB account.
+
+    .. code-block:: yaml
+
+        policies:
+          - name: store-on-hours-state
+            resource: azure.cosmosdb-collection
+            filters:
+              - type: offer
+                key: content.offerThroughput
+                op: gt
+                value: 400
+            actions:
+              - type: save-throughput-state
+                state-tag: on-hour-state
+
+    """
+
+    schema = type_schema(
+        'save-throughput-state',
+        required=['state-tag'],
+        **{
+            'state-tag': {'type': 'string'}
+        }
+    )
+
+    TAG_VALUE_CHAR_LIMIT = 256
+
+    def _process_resources(self, resources, event):
+        OfferHelper.execute_in_parallel_grouped_by_account(
+            resources,
+            self.executor_factory,
+            self.manager.get_parent_manager(),
+            self._process_account_set,
+            self.log
+        )
+
+    def _process_account_set(self, resources, account_client):
+        account_tag_values = []
+        tag_name = self.data.get('state-tag')
+        cosmos_account = resources[0]['c7n:parent']
+
+        for resource in resources:
+            account_tag_values.append('{}:{}'.format(
+                resource['_rid'], resource['c7n:offer']['content']['offerThroughput']))
+
+        tag_value = ';'.join(account_tag_values)
+
+        if len(tag_value) > self.TAG_VALUE_CHAR_LIMIT:
+            raise('Can not add tag, {}, on parent resource, {}, '
```

Comment:
Can we raise more specific ValidationError()? (I think this is some sort of runtime validation.. Open to other options)

Reply:
It can be ValueError.. ValidationError seems to be defined in jsonschema, so not the best candidate..